query/src/main/scala/filodb/query/exec/SetOperatorExec.scala

```diff
@@ -133,39 +155,72 @@ final case class SetOperatorExec(queryContext: QueryContext,
             rhsRows.close()
           }
         }
-        result += IteratorBackedRangeVector(lhs.key, rows)
+        val arrayBuffer = result.getOrElse(jk, ArrayBuffer())
+        val resRv = IteratorBackedRangeVector(lhs.key, rows)
+        if (index >= 0) arrayBuffer.update(index, resRv) else arrayBuffer.append(resRv)
+        result.put(jk, arrayBuffer)
       } else if (jk.isEmpty) {
         // "up AND ON (dummy) vector(1)" should be equivalent to up as there's no dummy label
-        result += lhs
+        val arrayBuffer = result.getOrElse(jk, ArrayBuffer())
+        arrayBuffer.append(lhs)
+        result.put(jk, arrayBuffer)
       }
     }
-    result.toList
+    result.values.flatten
```

Comment:
See if you can improve allocation by returning iterator of Range Vectors. Eventually, you can also create Observable from Iterator. This will reduce O(n) allocation.